feat: Introduce 'sanitizeKeys' config option

kamilogorek opened the pull request:
Resolves https://github.com/getsentry/raven-js/issues/296

It was quite requested feature by some of the clients recently, so we may get it in, in one form or another. I'm open for the feedback.

MaxBittker commented:
very cool! haven't reviewed the implementation in depth yet, but I think it's worth discussing the API/behavior/guarantees/terminology of this feature, because it could become the gold standard for our other clients going forward.

afaik (and I could be very wrong, @bretthoerner @HazAT ) the only clients which support this right now are python and ruby.


### Python
docs:
https://docs.sentry.io/clients/python/advanced/#sanitizing-data
implementation:
https://github.com/getsentry/raven-python/blob/master/raven/processors.py

* sanitizes a _fixed_ list of keys (AFAIK, not configurable by the user without making a new instance of the processor and passing that in)
* filters extra data and http data including query params

### Ruby:
docs:
https://docs.sentry.io/clients/ruby/config/#configuration
implementation:
https://github.com/getsentry/raven-ruby/blob/master/lib/raven/processor/sanitizedata.rb

* has some default patterns including password and credit card-like strings.
* also accepts a few configs including regexes and predicates I think?
* filters query strings & json blob

dcramer commented:
I'd prefer something simple like sanitizeKeys over the general processors pattern we implemented elsewhere. I think also supporting regexp is good.

kamilogorek commented:
The main issue I have with RegExp is that it can help people easily shoot themselves in the foot, as this sanitizer goes through *all* keys of the payload sent to Sentry. Including exception, message, frames, etc. etc.

MaxBittker commented on src/utils.js:
```diff
@@ -535,6 +535,30 @@ function serializeKeysForMessage(keys, maxLength) {
 }
 
+function sanitize(input, sanitizeKeys) {
+  sanitizeKeys = isArray(sanitizeKeys) ? sanitizeKeys : [];
```
if it's not an array, maybe just bail out here and return array? to reduce the risk involved in shipping this code?

ndmanvar commented:
> The main issue I have with RegExp is that it can help people easily shoot themselves in the foot, 
as this sanitizer goes through all keys of the payload sent to Sentry. Including exception, message, frames, etc. etc.

I can understand why this can be an example of "shooting yourself in the foot ", but many companies/organization want to scrub PII everywhere (frames, tags, extra information, error title/msg, etc.). Maybe a config that specifies to look through everything / etc.??

MaxBittker commented on src/utils.js:
```diff
@@ -535,6 +535,30 @@ function serializeKeysForMessage(keys, maxLength) {
 
+function sanitize(input, sanitizeKeys) {
+  sanitizeKeys = isArray(sanitizeKeys) ? sanitizeKeys : [];
+  var sanitizeMask = '********';
```
i know we use asterisks elsewhere, but i think something explicit like `[value sanitized by raven.js]`could be better.

probably not worth breaking with the precedent though

MaxBittker commented on src/utils.js:
```diff
@@ -535,6 +535,30 @@ function serializeKeysForMessage(keys, maxLength) {
   return '';
 }
 
+function sanitize(input, sanitizeKeys) {
```
how does this deal with cycles? or will the data be guaranteed by this point not to have any? how strong is that guarantee? (afraid of freezing a browser in some weird scenario)

kamilogorek replied:
It didn't. Thanks for catching that!

MaxBittker commented on docs/config.rst:
```diff
@@ -149,6 +149,17 @@ Those configuration options are documented below:
 
+.. describe:: sanitizeKeys
+
+    An array of strings representing keys that should be scrubbed from the payload sent to Sentry.
```
(worth being explicit here that this doesn't scrub or match on string values or query encoded data)

MaxBittker replied:
also worth mentioning that sentry itself can do do server side sanitizing and this is different

MaxBittker replied:
basically i think people are really gonna want to know what guarantees they get from this feature so it's worth writing a longer explanation

kamilogorek commented:
@dcramer added support for RegExp.
@MaxBittker handled cyclic references, updated docs and changed behavior to bail quickly with the input if incorrect keys type is used.
